insonmnia/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,351 @@
+	g.logger.Info("fix transaction",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()))
+
+	if !g.isCommitted(ctx, tx) {
+		_, err := g.out.Payout(ctx, g.key, tx.From, tx.Value, tx.Number)
+		if err != nil {
+			g.logger.Debug("error while commit", zap.Error(err))
+			return err
+		}
+		g.logger.Info("transaction committed",
+			zap.String("from", tx.From.String()),
+			zap.String("value", tx.Value.String()),
+			zap.String("tx number", tx.Number.String()))
+
+		// sleeping for freezing time after committing
+		time.Sleep(g.freezingTime)
+	}
+
+	if !g.transactionCommittedByMe(ctx, tx) {
+		return fmt.Errorf("transaction commited by other gate")
+	}
+
+	_, err := g.out.Payout(ctx, g.key, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		g.logger.Debug("error while payout", zap.Error(err))
+		return err
+	}
+	g.logger.Debug("transaction payouted",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()))
+
+	return nil
+}
+
+func (g *Gatekeeper) findScummyTransactions(ctx context.Context, inTxs map[string]*blockchain.GateTx, outTxs map[string]*blockchain.GateTx) {
+	for k, tx := range outTxs {
+		_, ok := inTxs[k]
+		if !ok {
+			go g.freezeScummy(ctx, tx)
+		}
+	}
+}
+
+func (g *Gatekeeper) freezeScummy(ctx context.Context, tx *blockchain.GateTx) error {
+	txState, err := g.out.GetTransactionState(ctx, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		return err
+	}
+
+	keeper, err := g.out.GetKeeper(ctx, txState.Keeper)
+	if err != nil {
+		return err
+	}
+
+	g.logger.Info("found new scum transaction",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()),
+		zap.Uint64("block number", tx.BlockNumber),
+		zap.String("commit timestamp", txState.CommitTS.String()),
+		zap.Bool("paid", txState.Paid),
+		zap.String("keeper address", keeper.Address.String()),
+		zap.String("keeper day limit", keeper.DayLimit.String()),
+		zap.String("keeper day limit", keeper.SpentToday.String()),
+		zap.String("keeper last day ", keeper.LastDay.String()),
+		zap.Bool("keeper frozen", keeper.Frozen))
+
+	if keeper.Frozen {
+		return fmt.Errorf("keeper already frozen")
+	}
+
+	return g.out.FreezeKeeper(ctx, g.key, txState.Keeper)
+}
+
+// loadFreezeTime watch current freezing time in contract
+func (g *Gatekeeper) loadFreezeTime(ctx context.Context) error {
+	var freezingTime *big.Int
+	var err error
+
+	freezingTime, err = g.out.GetFreezingTime(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !freezingTime.IsInt64() {
+		return fmt.Errorf("freezing time is abused")
+	}
+
+	// multiply to second because blockchain operate with time in seconds
+	ft := time.Duration(freezingTime.Int64()) * time.Second
+	if g.freezingTime == ft {
+		return nil
+	}
+	g.freezingTime = ft
```

Comment:
this can be racy, lock the mutex when operating with the shared variables into goroutines